pkg/sql/schemachanger/scpb/element_collection.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scpb
+
+import "github.com/cockroachdb/errors"
+
+// ElementCollection represents an ordered set of
+// (current status, target status, element) tuples,
+// looked up in `g` by iterating through `indexes`.
+type ElementCollection[E Element] struct {
+	g       ElementCollectionGetter
+	indexes []int
+}
+
+// ElementCollectionGetter abstracts the actual mechanism for obtaining
+// (current status, target status, element) tuples in an element collection.
+type ElementCollectionGetter interface {
+
+	// Get returns values for the given index in [0, Size()[.
+	Get(index int) (current Status, target TargetStatus, e Element)
+
+	// Size returns the cardinality of the collection.
+	Size() int
+}
+
+// NewElementCollection constructs a generic element collection
+// containing the elements in `g` listed in the `indexes` slice.
+// The slice is owned by the returned struct.
+func NewElementCollection(g ElementCollectionGetter, indexes []int) *ElementCollection[Element] {
+	if g == nil || len(indexes) == 0 {
+		return nil
+	}
+	return &ElementCollection[Element]{
+		g:       g,
+		indexes: indexes,
+	}
+}
+
+// ElementCollection is an ElementCollectionGetter.
+var _ ElementCollectionGetter = &ElementCollection[Element]{}
+
+// Get implements the ElementCollectionGetter interface.
+func (c *ElementCollection[E]) Get(ordinal int) (current Status, target TargetStatus, e Element) {
```

Comment:
Interesting. So `ElementCollection` is generic on `E` which is constrained to type `Element`. Here, the `Get()` method always return a type `Element`? Looks like generic is not needed here? Correct me if I'm wrong.

Reply:
`ElementCollection` has `Get` defined only because it implements `ElementCollectionGetter` which itself only exists to abstract away the object which actually stores these elements + states (it's `scbuild.builderState`)